flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java

```diff
@@ -252,28 +259,94 @@ public void testTransitionToFinishedOnSuspend() throws Exception {
+                    });
+            exec.onNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testNotifyNewResourcesAvailableAfterCooldownIsOverStateChange() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            final ExecutingStateBuilder executingStateBuilder =
+                    new ExecutingStateBuilder()
+                            .setScalingIntervalMin(Duration.ofSeconds(20L))
+                            .setLastRescale(Instant.now().minus(Duration.ofSeconds(30L)));
+            Executing exec = executingStateBuilder.build(ctx);
+            ctx.setCanScaleUp(true); // => rescale
+            ctx.setExpectRestarting(
+                    restartingArguments -> { // immediate rescale
+                        assertThat(restartingArguments.getBackoffTime(), is(Duration.ZERO));
+                        assertThat(ctx.actionWasScheduled, is(false));
+                    });
+            exec.onNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testNotifyNewResourcesAvailableWithCanScaleUpWithoutForceTransitionsToRestarting()
             throws Exception {
         try (MockExecutingContext ctx = new MockExecutingContext()) {
             Executing exec = new ExecutingStateBuilder().build(ctx);
 
             ctx.setExpectRestarting(
                     restartingArguments -> {
-                        // expect immediate restart on scale up
+                        // immediate rescale
                         assertThat(restartingArguments.getBackoffTime(), is(Duration.ZERO));
+                        assertThat(ctx.actionWasScheduled, is(false));
                     });
-            ctx.setCanScaleUp(() -> true);
+            ctx.setCanScaleUp(true); // => rescale
             exec.onNewResourcesAvailable();
         }
     }
 
     @Test
-    public void testNotifyNewResourcesAvailableWithNoResourcesAndNoStateChange() throws Exception {
+    public void testNotifyNewResourcesAvailableWithCantScaleUpWithoutForceAndCantScaleUpWithForce()
+            throws Exception {
         try (MockExecutingContext ctx = new MockExecutingContext()) {
-            Executing exec = new ExecutingStateBuilder().build(ctx);
-            ctx.setCanScaleUp(() -> false);
+            Executing exec =
+                    new ExecutingStateBuilder()
+                            .setScalingIntervalMax(Duration.ofSeconds(1L))
+                            .build(ctx);
+            ctx.setCanScaleUp(
+                    false, false); // => schedule force rescale but resource lost on timeout =>
+            // no rescale
             exec.onNewResourcesAvailable();
             ctx.assertNoStateTransition();
+            assertThat(ctx.actionWasScheduled, is(true));
+        }
+    }
+
+    @Test
+    public void testNotifyNewResourcesAvailableWithCantScaleUpWithoutForceAndCanScaleUpWithForce()
+            throws Exception {
```

Comment:
- if minimum scaling requirements are not met
  - case1: if last rescale was done more than scaling-interval.max ago, a rescale is forced.
  - case2: otherwise, schedule a forced rescale in scaling-interval.max

This test is case1, and the case2 isn't covered, right?

Reply:
yesterday we added an immediate force rescale if `timeSinceLastRescale > interval-max`. This case is not covered by the tests. The current test covers the previous behaviour in which we scheduled a force rescale no matter `timeSinceLastRescale`. I added a test for immediate force